### DISCUSSION TITLE
Guidance: Nested indexer [] within flow sequence

### DISCUSSION THREAD

**Discussion by evilrobot-01** [eemeli/yaml#423] [Q&A] [answered]
Is there any way to easily extend the parsing of flow sequence entries to allow values which contain the indexer [] characters? The requirement is simply to provide a string-type value in yaml, which is parsed in and ultimately passed to https://lodash.com/docs/#get to resolve a value within a particular object.

```
import YAML from 'yaml'

// Works when enclosed in quotes
YAML.parse('args: [ "$variable[0]", 100 ]')

// Required syntax
YAML.parse('args: [ $variable[0], 100 ]')
```

The second parse `YAML.parse` call results in:
```
YAMLParseError: Missing , or : between flow sequence items at line 1, column 18:

args: [ $variable[0], 100 ]
                 ^
```

**Comment by eemeli** [comment, marked as answer]
Short answer: not really, because the latter "required syntax" is not valid YAML.

Within flow collections, plain scalars cannot contain `[]` or `{}` characters: https://yaml.org/spec/1.2.2/#733-plain-style

They're fine is a block collection though:
```js
YAML.parse(`
args:
  - $variable[0]
  - 100
`)
// { args: [ '$variable[0]', 100 ] }
```

**Comment by evilrobot-01** [comment]
Thanks for the reply! I had figured out that the block collection works but the flow syntax is preferred.

Is there perhaps a way for me to add support somehow to meet a requirement, even if not technically valid yaml? Perhaps a custom tag could handle an entry when starting with $ or handle processing of args? 🙂
________________________________

**Reply by eemeli** [reply]
Not really, because this already happens in the lexer, and your input is ambiguous. Consider what happens here:

```
args: [ $variable[0], 100 ]
                   ^
```

If `[]` were allowed in scalar flow values, how could we know at that point whether the outer `[`...`]` collection was being closed or not?

**Comment by evilrobot-01** [comment]
I see, thank you.

It seems the only way then would be if the lexer was somehow tolerant of nested [ ], perhaps by using a stack to track corresponding opening and closing brackets, but as you point out is invalid yaml.
________________________________
